zeppelin-server/src/test/java/org/apache/zeppelin/service/InterpreterServiceTest.java

```diff
@@ -0,0 +1,133 @@
+
+@RunWith(MockitoJUnitRunner.class)
+public class InterpreterServiceTest {
+  @Mock ZeppelinConfiguration mockZeppelinConfiguration;
+  @Mock NotebookServer mockNotebookServer;
+  @Mock InterpreterSettingManager mockInterpreterSettingManager;
+
+  Path temporaryDir;
+  Path interpreterDir;
+  Path localRepoDir;
+
+  InterpreterService interpreterService;
+
+  @Before
+  public void setUp() throws Exception {
+    temporaryDir = Files.createTempDirectory("tmp");
+    interpreterDir = Files.createTempDirectory(temporaryDir, "interpreter");
+    localRepoDir = Files.createTempDirectory(temporaryDir, "local-repo");
+
+    when(mockZeppelinConfiguration.getInterpreterDir()).thenReturn(interpreterDir.toString());
+    when(mockZeppelinConfiguration.getInterpreterLocalRepoPath())
+        .thenReturn(localRepoDir.toString());
+
+    when(mockZeppelinConfiguration.getZeppelinProxyUrl()).thenReturn(null);
+    when(mockZeppelinConfiguration.getZeppelinProxyUser()).thenReturn(null);
+    when(mockZeppelinConfiguration.getZeppelinProxyPassword()).thenReturn(null);
+
+    interpreterService =
+        new InterpreterService(mockZeppelinConfiguration, mockInterpreterSettingManager);
+  }
+
+  @After
+  public void tearDown() throws Exception {
+    if (null != temporaryDir) {
+      FileUtils.deleteDirectory(temporaryDir.toFile());
+    }
+  }
+
+  @Test(expected = Exception.class)
+  public void invalidProxyUrl() throws Exception {
+    when(mockZeppelinConfiguration.getZeppelinProxyUrl()).thenReturn("invalidProxyPath");
+
+    interpreterService.installInterpreter(
+        new InterpreterInstallationRequest("name", "artifact"), null);
+  }
+
+  @Test(expected = Exception.class)
+  public void interpreterAlreadyExist() throws Exception {
+    String alreadyExistName = "aen";
+    Path specificInterpreterDir =
+        Files.createDirectory(Paths.get(interpreterDir.toString(), alreadyExistName));
+
+    interpreterService.installInterpreter(
+        new InterpreterInstallationRequest(alreadyExistName, "artifact"), null);
+  }
+
+  @Test(expected = Exception.class)
+  public void interpreterAlreadyExistWithDifferentName() throws Exception {
+    String interpreterName = "in";
+    Files.createDirectory(Paths.get(interpreterDir.toString(), interpreterName));
+
+    String anotherButSameInterpreterName = "zeppelin-" + interpreterName;
+
+    interpreterService.installInterpreter(
+        new InterpreterInstallationRequest(anotherButSameInterpreterName, "artifact"), null);
+  }
+
+  @Test
+  public void downloadInterpreter() throws IOException {
+    String interpreterName = "test-interpreter";
+    String artifactName = "junit:junit:4.11";
+    Path specificInterpreterPath =
+        Files.createDirectory(Paths.get(interpreterDir.toString(), interpreterName));
+    DependencyResolver dependencyResolver = new DependencyResolver(localRepoDir.toString());
+
+    doNothing().when(mockInterpreterSettingManager).refreshInterpreterTemplates();
```

Comment:
Need to verify that refreshInterpreterTemplates is called when download is succeeded

Reply:
Let me try to solve it :-)

Reply:
Changed test case to check callback's message